Added tests and no arch

alen12345 opened the pull request.

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe```) and found it was in an excellent condition.

TimShawver commented:
Cool thanks.  Using `noarch` sounds fine.  Looks like the `circleci` builds failed due to not being able to find the tests? Your changes look right to me so it's not clear to me why the tests weren't found.

alen12345 commented:
@TimShawver I'll have a look on Friday or Saturday and see why they are not found ;)

isuruf commented:
You need to copy the test files to the test dir using https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#source-files
Also, noarch is not possible for recipes with `post-link` scripts

alen12345 commented:
@isuruf I did not think about that for the noarch part; I will also think about the chance to uniform the post installation operations on all platforms so that we can switch to no arch.

alen12345 commented:
@TimShawver I am wondering if we actually need the PREFIX part [here](https://github.com/conda-forge/qgrid-feedstock/blob/991da160a87d447b30ec5b72791ca32b08d7779e/recipe/post-link.sh#L1) and by removing the dev null we could make it no arch... need to try when I have more time....

TimShawver commented:
Cool, looks like the `bqplot` feedstock is no longer sending output to `dev/null` and instead is sending to `"${PREFIX}/.messages.txt"`.  See https://github.com/conda-forge/bqplot-feedstock/blob/master/recipe/post-link.sh.  So maybe we could send to a file as well.

I don't know if we actually need the `${PREFIX}` part, I got that from the `bqplot` build.  Looking at the conda docs though seems like that's the install path of the conda packages...I'm wondering if we might need that to figure out where `/bin/jupyter-nbextension`  could be found.

alen12345 commented:
@TimShawver I will experiment a bit and see what we can achieve :)

alen12345 commented:
@TimShawver this is what I have done:

-  removed all the pre and post scripts
- noarch by including those commands in the build command
- added tests

But, the tests in the upstream are failing for python2, since on line 44 we have the nonlocal keyword. We need to fix the test in the upstream unfortunately. Shall we have a look at it together?

alen12345 commented:
@TimShawver this to me looks like a good [solution](https://technotroph.wordpress.com/2012/10/01/python-closures-and-the-python-2-7-nonlocal-solution/). Can we please implement it in the upstream and then report back here so that we can change package version to 1.0.2 and merge the new recipe?

TimShawver commented:
Cool, thanks for working on this! I created a new beta version which we can use for testing the build.  I didn't want to release an official version yet in case we discover something else we need to change.

To try the new version, you can merge my PR here: https://github.com/alen12345/qgrid-feedstock/pull/1.

Have you tried installing locally to confirm that the post-build step is completing successfully? I'll do some test installs myself before we merge, but I just was curious about which platforms you've tested on already.

TimShawver commented on recipe/meta.yaml:
```diff
@@ -11,24 +11,35 @@ source:
+  number: 2
+  noarch: python
+  # disable if already installed
+  script: (jupyter-nbextension disable qgrid --py --sys-prefix || true) && python -m pip install --no-deps --ignore-installed . && jupyter-nbextension enable qgrid --py --sys-prefix
```
This does not appear to work for enabling qgrid automatically...I tried building this recipe locally and qgrid was not enabled after installing from the generated conda package (and it was enabled when I tried with a package that was built from the recipe on master).  

Looking at this again it looks like this `script:` option is in the `build:` section which means this will run as part of the build script on our CI services (travis, etc).  I don't think that's what we want because we need the `jupyter-nbextension enable` command to run on each qgrid user's machine after they install qgrid.

Unless you have something else you want to try, maybe we can just revert the `noarch` portion of this and keep the automated testing portion?

TimShawver commented:
@alen12345 @isuruf @jakirkham does anyone know why I'm seeing this `noarch` version of qgrid, which hasn't been merged to master yet, listed as the latest version of qgrid in the conda-forge channel?

![screen shot 2018-03-03 at 1 01 52 pm](https://user-images.githubusercontent.com/5055866/36937574-695097fe-1ee3-11e8-97bb-0f7a67e94bb2.png)

My assumption was that new versions of the conda-forge package would only be published once we merge PRs to master, yet somehow a version of qgrid which was seemingly built from a recipe on a fork, is listed as the latest version of qgrid, and is being installed on people's machine when they type `conda install qgrid`.  This is likely going to cause problems for qgrid users so I could use some help understanding why it happened and how to avoid it in the future. Thanks.

isuruf commented:
It's because somebody pushed it to a branch called `noarch-test-tim`

TimShawver commented:
Gotcha...sorry that was me.  I realized my mistake, then moved my PR to a fork.  I didn't realize pushing a branch would cause a new version to be published.  

If you could delete it (looks like I don't have permission) that would be great. Otherwise probably not a big deal because I'll be releasing a new version soon, once the builds complete for this PR: https://github.com/conda-forge/qgrid-feedstock/pull/5

isuruf commented:
No problem. I've moved the noarch package to broken label
